components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/PingScheduler.kt

```diff
@@ -0,0 +1,206 @@
+
+    // The application id detected by glean to be used as part of the submission
+    // endpoint.
+    internal var applicationId = sanitizeApplicationId(applicationContext.packageName)
+
+    companion object {
+        /**
+         * Helper function to check to see if a particular ping has been scheduled by the PingScheduler
+         *
+         * @return True if [PingUploadWorker] task found in [WorkManager], false otherwise
+         */
+        internal fun isWorkScheduled(): Boolean {
+            val instance = WorkManager.getInstance()
+            val statuses = instance.getWorkInfosByTag(PingUploadWorker.PING_WORKER_TAG)
+            return try {
+                var running = false
+                val workInfoList = statuses.get()
+                for (workInfo in workInfoList) {
+                    val state = workInfo.state
+                    running = (state === WorkInfo.State.RUNNING) or (state === WorkInfo.State.ENQUEUED)
+                }
+                running
+            } catch (e: ExecutionException) {
+                // Do nothing but return false
+                false
+            } catch (e: InterruptedException) {
+                // Do nothing but return false
+                false
+            }
+        }
+    }
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    /**
+     * Handles necessary glean tasks upon entering the background state [Lifecycle.Event.ON_STOP]
+     */
+    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+    fun onEnterBackground() {
+        // We're going to background, so store how much time we spent
+        // on foreground.
+        GleanBaseline.duration.stopAndSum()
+
+        if (!Glean.isInitialized()) {
+            logger.error("Glean not initialized, abort serialize and send ping")
+            return
+        }
+
+        // Do not send ping if upload is disabled
+        if (!Glean.getUploadEnabled()) {
+            logger.debug("Glean upload not enabled, abort serialize and send ping.")
+            return
+        }
+
+        // Schedule the baseline and event pings and save the result to determine whether or not
+        // we need to schedule the upload worker
+        var pingsToSchedule = serializePing(BaselinePing.STORE_NAME)
+        pingsToSchedule = pingsToSchedule || serializePing("events")
+
+        // Check the metrics ping schedule to determine whether it's time to schedule a new
+        // metrics ping or not
+        if (metricsPingScheduler.canSendPing()) {
+            pingsToSchedule = pingsToSchedule || serializePing(MetricsPingScheduler.STORE_NAME)
+            metricsPingScheduler.updateSentTimestamp()
+        }
+
+        // It should only take a single PingUploadWorker to process all queued pings, so we only
+        // want to have one scheduled at a time.
+        if (pingsToSchedule && !isWorkScheduled()) {
+            WorkManager.getInstance().enqueue(buildWorkRequest())
```

Comment:
I think we need to make sure that this happens *after* pings are written to disk. Otherwise, we might have some delay in the ping submission. Think of this case:

- we assemble ping A and async write it to disk
- we enqueue/start the work request but A wasn't written to disk yet
- the worker goes through the files, doesn't find anything and bails out
- the worker is enqueued again next time a ping is sent (which could be a bit away in time)

Let's *not* address this problem in this bug. Let's file a bug, add a comment here with the bug number, and address it later, if required by data quality analysis.

Reply:
Simple enough, will file bug about it.